### DISCUSSION TITLE
ClassificationTask example

### DISCUSSION THREAD

**Discussion by robmarkcole** [torchgeo/torchgeo#789] [Q&A]
Hi
I wish to create an example notebook training a ClassificationTask on the Merced dataset using pytorch lightning, and adapted the approach in https://torchgeo.readthedocs.io/en/latest/tutorials/trainers.html without success.

I have:

```python
datamodule = UCMercedDataModule(
     root_dir = data_dir, batch_size=4, num_workers=2
)

task = ClassificationTask(
    classification_model="resnet18",
    weights="imagenet",
    num_classes=num_classes,
    in_channels=3,
    loss="ce", 
    learning_rate=0.1,
    learning_rate_schedule_patience=5
)

experiment_dir = os.path.join(tempfile.gettempdir(), "merced_results")

checkpoint_callback = ModelCheckpoint(
    monitor="val_loss",
    dirpath=experiment_dir,
    save_top_k=1,
    save_last=True,
)

early_stopping_callback = EarlyStopping(
    monitor="val_loss",
    min_delta=0.00,
    patience=10,
)

csv_logger = CSVLogger(
    save_dir=experiment_dir,
    name="tutorial_logs"
)

trainer = pl.Trainer(
    callbacks=[checkpoint_callback, early_stopping_callback],
    logger=[csv_logger],
    default_root_dir=experiment_dir,
    min_epochs=1,
    max_epochs=100,
    gpus=1
)

trainer.fit(model=task, datamodule=datamodule)
```

However training proceeds for only 1 epoch then exits.  Can anyone steer me in the right direction?

**Comment by adamjstewart** [comment]
What happens when you change `min_epochs` from 1 to 10? I'm not actually sure how early stopping would work if you only have a single epoch in the history to compare to.

**Reply by robmarkcole** [reply]
Strange, reran today with those params and training proceeds

**Comment by isaaccorley** [comment]
I was able to train and test fine for 50 epochs with the following code and installing the latest torchgeo v0.3.1 via `!pip install torchgeo` in a google colab gpu instance. Here's a [link](https://gist.github.com/isaaccorley/17f9ce96eb075550800d0c27bc9d8fe1) to the colab notebook.


```python
import pytorch_lightning as pl

from torchgeo.datasets import UCMerced
from torchgeo.datamodules import UCMercedDataModule
from torchgeo.trainers import ClassificationTask


# Parameters
data_dir = "./data"
num_classes = 21
channels = 3
batch_size = 4
num_workers = 2
backbone = "resnet18"
weights = "imagenet"
lr = 0.01
lr_schedule_patience = 5
epochs = 50

# Download dataset and dataset splits
dataset = UCMerced(data_dir, download=True, checksum=True)

# Instantiate datamodule, classifier task, and callbacks
datamodule = UCMercedDataModule(
     root_dir=data_dir,
     batch_size=batch_size,
     num_workers=num_workers,
)
task = ClassificationTask(
    classification_model=backbone,
    weights=weights,
    num_classes=num_classes,
    in_channels=channels,
    loss="ce", 
    learning_rate=lr,
    learning_rate_schedule_patience=lr_schedule_patience
)
checkpoint_callback = pl.callbacks.ModelCheckpoint(
    monitor="val_loss",
    save_top_k=1,
    save_last=True,
)
early_stopping_callback = pl.callbacks.EarlyStopping(
    monitor="val_loss",
    min_delta=0.00,
    patience=10,
)

# Train
trainer = pl.Trainer(
    callbacks=[checkpoint_callback, early_stopping_callback],
    max_epochs=epochs,
    gpus=1
)
trainer.fit(model=task, datamodule=datamodule)

# Test
test_metrics = trainer.test(model=task, datamodule=datamodule)

"""
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       Test metric             DataLoader 0
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  test_AverageAccuracy      0.7456777691841125
      test_F1Score          0.7523809671401978
    test_JaccardIndex       0.6187251210212708
  test_OverallAccuracy      0.7523809671401978
        test_loss           0.9027961492538452
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
"""
```

Here are the tensorboard train/val loss and overall acc plots over steps:

![image](https://user-images.githubusercontent.com/22203655/192051779-3b2a9c2d-4acf-49f5-a9a3-c7582518dc40.png)

![image](https://user-images.githubusercontent.com/22203655/192051798-122684a0-149a-43c2-9ca9-b3c816141b69.png)

![image](https://user-images.githubusercontent.com/22203655/192051657-5f068f72-7b9c-4227-997c-a6e1fd7c3cdd.png)

![image](https://user-images.githubusercontent.com/22203655/192051696-f2f9ff9c-8f5c-432c-8f1b-4b0226a1041b.png)

**Reply by calebrob6** [reply]
🤔 🙂 

| **Method**                            	| **UCMerced Test Overall Accuracy** 	|
|---------------------------------------	|:----------------------------------:	|
| ResNet50 v2 with InDomain pretraining 	|               99.61%               	|
| ResNet50 as reported in TorchGeo      	|          98.15% +/- 0.46%          	|
| Caleb (Github, 2022)                  	|               98.09%               	|
| Isaac (Github, 2022)                  	|               75.24%               	|

**Reply by isaaccorley** [reply]
***changes learning rate > gets slightly better acc > publishes to MDPI Remote Sensing***

**Reply by calebrob6** [reply]
Change the preprocess method to resize to 224 by 224 and use a swin transformer

**Reply by calebrob6** [reply]
In seriousness though @robmarkcole, you _can_ use any `timm` model in `torchgeo.trainers.ClassificationTask`. E.g. I just re-ran with a "resnext50_32x4d" for kicks and got 98.81%. While this isn't super interesting on UCMerced, we've previously discussed that it would be interesting to flush out the performance difference between the basic recipes that work (e.g. ResNet18s here, and more generally, the results we report in the torchgeo paper) and SOTA methods.

**Reply by isaaccorley** [reply]
Swin Transformer results in `96.4% OA`with `swin_base_patch4_window7_224` and resizing to `224x224`. Just use any timm model instead of `resnet18` in the classification task.

**Reply by robmarkcole** [reply]
Thanks for you effort on this guys. I didn't realise any timm model can be used, that is fantastic. My plan is for a 'getting started with classification' blog post, and including a section on how you can go from basic model to publication level results would be very interesting. Could cross post on personal blog and TorchGeo blog?

**Reply by adamjstewart** [reply]
We would love to add a new Jupyter Notebook tutorial on that! We don't yet have a blog platform other than the original blog we published at pytorch.org/blog but we could consider doing something on Medium.

**Reply by robmarkcole** [reply]
@adamjstewart I am happy to work on the notebook and contribute a PR. Rough idea is then: 
- intro basic resnet18 model
- resize to 224 by 224 and use a swin transformer, show improvement

In [my blog](https://robmarkcole.com/) post I would then just reference this notebook

**Comment by calebrob6** [comment]
Aw man, @isaaccorley beat me to it by a few minutes, but I crushed his score :)...

Here's mine (lower LR is better) -- https://gist.github.com/calebrob6/ebfbf202977b2f8f27409ddeaac479d2

```
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       Test metric             DataLoader 0
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  test_AverageAccuracy      0.9774090647697449
      test_F1Score          0.9809523820877075
    test_JaccardIndex       0.9600954055786133
  test_OverallAccuracy      0.9809523820877075
        test_loss           0.09108178317546844
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

![image](https://user-images.githubusercontent.com/1088078/192056310-ab7624bb-6467-413b-acce-e0f6f06c95f6.png)

**Reply by isaaccorley** [reply]
Good point on lower learning rate being better. I initially ran with `lr=0.1` in OP's code and only got around 30% OA.
